### PR TITLE
grafana-loki: 2.2.1 -> 2.3.0

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

upstream update
https://github.com/grafana/loki/releases/tag/v2.3.0
the license changed.
Leaving this a couple of days open in case anybody wants to check it out.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 133011` at b05113e5 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/6b438cb29a476a11b853d7bf84b70de8">grafana-loki</a></li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/servers/monitoring/loki/default.nix [hunk 1]
```diff
@@ -47,7 +47,7 @@ buildGoModule rec {
 
   meta = with lib; {
     description = "Like Prometheus, but for logs";
-    license = licenses.asl20;
+    license = licenses.agpl3Only;
```
Comment (mweinelt) [inline pkgs/servers/monitoring/loki/default.nix:50]:
```suggestion
    license = with licenses; [ 
      asl20
      agpl3Only
    ];
```

**Comment by mweinelt** [comment]
Marked as security, because of a vulnerability in cortex.

> Restrict path segments in TenantIDs (CVE-2021-36156 CVE-2021-36157).

And as the bot noted, this does not build at all.

**Review by amaxine — CHANGES_REQUESTED** [review]
`2.3.0` requires go 1.16, once it's unpinned in top-level it'll build. `io/fs` was only introduced in go 1.16

Additionally promtail moved from `cmd/promtail` to `clients/cmd/promtail`. That also needs to be reflected in this bump.

It also seems that version output is a bit subpar now (perhaps due to the recent changes to buildGoModule?). Would be nice to get that fixed too :)
```
loki, version  (branch: , revision: )
  build user:       
  build date:       
  go version:       go1.16.7
  platform:         linux/amd64
```

Forgot to add, I didn't actually test if everything works with those changes, just made sure it builds and looked at what `loki version` returns.

**Comment by happysalada** [comment]
@mweinelt @maxeaubrey 
This is some quality review! Thank you!
I've done the changes pointed out.
I've pinned go to the 1.16 version, just so that when they want to upgrade the default go modules, it doesn't affect this package unintendedly.

The loki version command has been restored to what it was. (it was just a path change).

File: pkgs/servers/monitoring/loki/default.nix [hunk 2]
```diff
@@ -38,16 +38,27 @@ buildGoModule rec {
...
+      "${t}.Version=${version}"
+      "-X"
+      "${t}.BuildUser=nix@nixpkgs"
+      "-X"
+      "${t}.BuildDate=unknown"
+      "-X"
+      "${t}.Branch=unknown"
+      "-X"
+      "${t}.Revision=unknown"
+    ];
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/loki/default.nix:55]:
```suggestion
  ldflags = let t = "github.com/grafana/loki/pkg/util/build"; in [
    "-s"
    "-w"
    "-X ${t}.Version=${version}"
    "-X ${t}.BuildUser=nix@nixpkgs"
    "-X ${t}.BuildDate=unknown"
    "-X ${t}.Branch=unknown"
    "-X ${t}.Revision=unknown"
  ];
```

**Pull request merged**